mmv1/third_party/terraform/utils/bootstrap_utils_test.go

```diff
@@ -516,6 +516,77 @@ func BootstrapProject(t *testing.T, projectID, billingAccount string, services [
 	return project
 }
 
+// BootstrapPSARolesOptions are for bootstrapping service agent roles in a project.
+type BootstrapPSARolesOptions struct {
+	ProjectID string // the project in which to grant roles
+	AgentName string // the name of the service agent
+	// (e.g. for service-<project_number>@gcp-sa-healthcare.iam.gserviceaccount.com,
+	// agentName would be 'gcp-sa-healthcare')
+	Roles []string // the roles to grant
+}
+
+// BootstrapPSARoles ensures that the given project's IAM
+// policy grants the given service's agent the given roles.
+// This is important to bootstrap because using iam policy resources means that
+// deleting removes permissions for concurrent tests.
+// Return the roles granted.
+func BootstrapPSARoles(t *testing.T, opts BootstrapPSARolesOptions) []string {
+	config := BootstrapConfig(t)
+	if config == nil {
+		t.Fatal("Could not bootstrap a config for BootstrapPSARoles.")
+	}
+	client := config.NewResourceManagerClient(config.userAgent)
+
+	project, err := client.Projects.Get(opts.ProjectID).Do()
+	if err != nil {
+		t.Fatalf("Error getting project with id %q: %s", opts.ProjectID, err)
+	}
+
+	serviceAgent := fmt.Sprintf("serviceAccount:service-%d@%s.iam.gserviceaccount.com", project.ProjectNumber, opts.AgentName)
+	getPolicyRequest := &cloudresourcemanager.GetIamPolicyRequest{}
+	policy, err := client.Projects.GetIamPolicy(project.ProjectId, getPolicyRequest).Do()
+	if err != nil {
+		t.Fatalf("Error getting project iam policy: %v", err)
+	}
+
+	// Map each role to whether it was already in the policy
+	rolesFound := make(map[string]bool, len(opts.Roles))
```

Comment:
Did you consider some of the iam utility functions we have available, like `mergeBindings`, and `compareBindings` to detect changes? While probably slower technically, those functions have unit tests that might make them a bit safer. This bootstrap function will be modifying our root project policy, so any unexpected behavior could have wider impact.

Reply:
I did not realize we had those. I've refactored to make use of them.